Toolkit/ArcGISToolkit/BookmarksTableViewController.swift

```diff
@@ -0,0 +1,168 @@
+/// pan/zoom the geoView to the bookmark's viewpoint.  Users can change this default behavior by supplying a custom handler to the `bookmarkSelectedHandler` property.
+/// When created using an array of `AGSBookmarks`, there is no default `bookmarkSelectedHandler` and the user must provide their own.
+public class BookmarksTableViewController: UITableViewController {
+    public typealias BookmarkSelectedFunction = (AGSBookmark) -> Void
+
+    /// The `AGSGeoView` containing either an `AGSMap` or `AGSScene` with the `AGSBookmark`s to display.
+    /// - Since: 100.7.0
+    public var geoView: AGSGeoView? {
+        didSet {
+            if let mapView = geoView as? AGSMapView {
+                mapView.map?.load { [weak self] (error) in
+                    if let error = error {
+                        print("Error loading map: \(error)")
+                    }
+                    self?.bookmarks = mapView.map?.bookmarks as? [AGSBookmark] ?? []
+                }
+                
+                // Add an observer to handle changes to the map.bookmarks array.
+                bookmarksObservation = mapView.map?.observe(\.bookmarks, options: [.new, .old]) { [weak self] (_, _) in
+                    self?.bookmarks = mapView.map?.bookmarks as? [AGSBookmark] ?? []
+                }
+                
+                // Add an observer to handle changes to the mapView.map.
+                mapOrSceneObservation = mapView.observe(\.map, options: [.new, .old]) { [weak self] (_, _) in
+                    self?.bookmarks = mapView.map?.bookmarks as? [AGSBookmark] ?? []
+                }
+            } else if let sceneView = geoView as? AGSSceneView {
+                sceneView.scene?.load { [weak self] (error) in
+                    if let error = error {
+                        print("Error loading scene: \(error)")
+                    }
+                    self?.bookmarks = sceneView.scene?.bookmarks as? [AGSBookmark] ?? []
+                }
+                
+                // Add an observer to handle changesto the scene.bookmarks array.
+                bookmarksObservation = sceneView.scene?.observe(\.bookmarks, options: [.new, .old]) { [weak self] (_, _) in
+                    self?.bookmarks = sceneView.scene?.bookmarks as? [AGSBookmark] ?? []
+                }
+                
+                // Add an observer to handle changes to the sceneView.scene.
+                mapOrSceneObservation = sceneView.observe(\.scene, options: [.new, .old]) { [weak self] (_, _) in
+                    self?.bookmarks = sceneView.scene?.bookmarks as? [AGSBookmark] ?? []
+                }
+            }
+        }
+    }
+        
+    /// The array of `AGSBookmark`s to display.
+    /// - Since: 100.7.0
+    public var bookmarks = [AGSBookmark]() {
+        didSet {
+            DispatchQueue.main.async { [weak self] in
```

Comment:
Not quite sure what you mean here...

Reply:
I mean that users are most likely to set this from the main thread. So rather than asynchronously dispatching to the main thread all the time (even when we're already on the main thread), you might consider dispatching only when needed, in the property observations:

```swift
bookmarksObservation = mapView.map?.observe(\.bookmarks) { [weak self] (map, _) in
    setBookmarksOnMainThread(map.bookmarks as? [AGSBookmark] ?? [])
}

public var bookmarks = [AGSBookmark]() {
    didSet {
        tableViewController.bookmarks = self.bookmarks
    }
}
```

But actually, the geo model is a property on a view itself, so really it should only be accessed from the main thread as well. In that case, it may be best to just dispatch inside the observation:

```swift
bookmarksObservation = mapView.map?.observe(\.bookmarks) { [weak self] (map, _) in
    DispatchQueue.main.async {
        self?.bookmarks = map.bookmarks as? [AGSBookmark] ?? []
    }
}
```